File: mysterious.py
# Problem: Mysterious Present

# Chain here is such a sequence of envelopes A = {a1,  a2,  ...,  an}, where the width and 
# the height of the i-th envelope is strictly higher than the width and the height of the 
# (i  -  1)-th envelope respectively. Chain size is the number of envelopes in the chain.

# We have to make the chain of the maximum size from the envelopes, such that we will be able
# to put a card into it. The card fits into the chain if its width and height is lower than the 
# width and the height of the smallest envelope in the chain respectively.


import sys

# Variables

n=0 # number of envelopes
w=0	# card width
h=0	# card height

# Parse input

flag=0
count=0

chain=[]
for line in sys.stdin:
	if flag==0:		# first line			
		n=int(line.split()[0])
		w=int(line.split()[1])
		h=int(line.split()[2])
		flag+=1
	else:	
		    		# i-th line
		count+=1
		width=int(line.split()[0])
		height=int(line.split()[1])

		if width>w and height>h:  # check if card fits in this envelope

			tup=((width,height),count)	# tuple of dimensions and index
			chain.append(tup)

		if count==n:
			break



if len(chain)==0: # card does not fit in any envelope
	print(0)

else:

	# now sort the chain of envelopes
	chain.sort()

	# now we have to use Recursion or Dynamic programming check through every possible branch 
	# to get the longest chain possible

	n = len(chain)
	dp = [0]*n
	parent = [-1]*n
  
	# dp is the list that stores longest increasing subsequence
	# Declare the list (array) for dp and initialize dp 
	# values for all indexes 

	dp[0]=1       # initialize first element of dp
	for i in range(1,n):
		for j in range(0,i):
			if chain[j][0][0] < chain[i][0][0] and chain[j][0][1]<chain[i][0][1]:
				if 1+dp[j] > dp[i]:
					dp[i]=1+dp[j]
					parent[i]=j

		if(dp[i]==0):
			dp[i]=1


	# Initialize maximum to 0 to get the maximum of all 
	maximum = 1
	k=0
  
	# Pick maximum of all values 
	for i in range(n):
		if dp[i]>maximum:
			maximum = dp[i]
			k=i

	print(maximum)

	final_indexes=[]
	while True:
		final_indexes.append(chain[k][1]) # stored in reverse order
		k=parent[k]
		if k==-1:
			break

	for item in reversed(final_indexes):
		print(item)

File: mysterious.cpp
/*

Chain here is such a sequence of envelopes A = {a1,  a2,  ...,  an}, where the width and 
the height of the i-th envelope is strictly higher than the width and the height of the 
(i  -  1)th envelope respectively. Chain size is the number of envelopes in the chain.

We have to make the chain of the maximum size from the envelopes, such that we will be able
to put a card into it. The card fits into the chain if its width and height is lower than the 
width and the height of the smallest envelope in the chain respectively.

author: https://github.com/Waqar-107 

*/

#define _CRT_SECURE_NO_WARNINGS

#include<algorithm>
#include<cctype>
#include<cmath>
#include<cstdio>
#include<cstdlib>
#include<cstring>
#include<functional>
#include<iomanip>
#include<iostream>
#include<list>
#include<map>
#include<numeric>
#include<queue>
#include<set>
#include<stack>
#include<string>
#include<utility>
#include<vector>

#define N 5010

using namespace std;

struct env
{
	int w, h, idx;
	env() {}
	env(int w, int h, int idx) {
		this->w = w;
		this->h = h;
		this->idx = idx;
	}
};

bool cmp(env a, env b) {
	if (a.w == b.w)
		return a.h < b.h;

	return a.w < b.w;
}

int w, h;
vector<env> vec;
int dp[N];
int parent[N];

int main()
{
	ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	
  	int i, j, k;
	int n, m;

	cin>>n>>w>>h;

	for (i = 1; i <= n; i++)
	{

		cin>>j>>k;
		if (j > w && k > h)
			vec.push_back(env(j, k, i));
	}

	if (vec.size() == 0) {
		cout<<"0"<<endl;
		return 0;
	}

	sort(vec.begin(), vec.end(), cmp);

	n = vec.size();

	memset(dp, 0, sizeof(dp));
	memset(parent, -1, sizeof(parent));

	dp[0] = 1;
	for (i = 1; i < n; i++)
	{
		for (j = i - 1; j >= 0; j--)
		{
			if (vec[j].w < vec[i].w && vec[j].h < vec[i].h)
			{
				if (1 + dp[j] > dp[i])
				{
					dp[i] = 1 + dp[j];
					parent[i] = j;
				}
			}
		}

		if (!dp[i])dp[i] = 1;
	}

	int mx = 1; k = 0;
	for (i = 1; i < n; i++)
	{
		if (dp[i] > mx)
			mx = dp[i], k = i;
	}


	cout<<mx<<endl;

	vector<int> ans; ans.push_back(vec[k].idx);
	while (1)
	{
		k = parent[k];
		if (k == -1)break;

		ans.push_back(vec[k].idx);
	}

	reverse(ans.begin(), ans.end());
	for (int e : ans)
		cout<<e<<endl;

	return 0;
}
